File: code/ch06/6.4.1.sqrtx.py
class Solution:
    def mySqrt(self, x: int) -> int:
        l, h = 0, x
        while l <= h:
            mid = l + (h - l) // 2
            if l == h or l + 1 == h:
                break
            elif mid * mid > x:
                h = mid - 1
            else:
                l = mid
        if h * h <= x:
            return h
        else:
            return l

File: code/ch06/6.4.1.sqrtx.cpp
class Solution {
public:
    int mySqrt(int x) {
        long l = 0, h = x;
        while (l <= h) {
            long mid = l + (h - l) / 2;
            if (l == h || l + 1 == h) {
                break;
            } else if (mid * mid > x) {
                h = mid - 1;
            } else {
                l = mid;
            }
        }
        if (h * h <= x) {
            return (int)h;
        } else {
            return (int)l;
        }
    }
};